File: uhrwerk-config/src/main/java/io/qimia/uhrwerk/config/representation/Transform.java
package io.qimia.uhrwerk.config.representation;

import io.qimia.uhrwerk.config.ConfigException;

import java.util.Arrays;

public class Transform{

    private String type;
    private Partition partition;

    public Transform() {}

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public Partition getPartition() {
        return partition;
    }

    public void setPartition(Partition partition) {
        this.partition = partition;
    }

    public void validate(String path){
        path += "transform/";
        if(type == null){
            throw new ConfigException("Missing field: " + path + "type");
        }
        if (!Arrays.asList("identity", "aggregate", "window", "temporal_aggregate").contains(type)) {
            throw new ConfigException("Wrong type! '" + type + "' is not allowed in " + path + "type");
        }
        if(!type.equals("identity")){
            if(partition == null){
                throw new ConfigException("Missing field: " + path + "partition");
            }
            else{
                partition.validate(path, type);
            }
        }
    }

    @Override
    public String toString() {
        return "Transform{" +
                "type='" + type + '\'' +
                ", partition=" + partition +
                '}';
    }
}

File: uhrwerk-config/src/main/kotlin/io/qimia/uhrwerk/config/representation/Transform.kt
package io.qimia.uhrwerk.config.representation

import io.qimia.uhrwerk.config.builders.ConfigException

data class Transform(

    var type: TransformType? = TransformType.NONE,
    var partition: Partition? = null
) {

    fun validate(path: String) {
        var path = path
        path += "transform/"
        if (type == null) {
            throw ConfigException("Missing field: " + path + "type")
        }
        if (!TransformType.values().contains(type)) {
            throw ConfigException("Wrong type! '" + type + "' is not allowed in " + path + "type")
        }
        if (type != TransformType.IDENTITY) {
            if (partition == null) {
                throw ConfigException("Missing field: " + path + "partition")
            } else {
                partition!!.validate(path, type!!.name)
            }
        }
    }

}